kubelet/rkt: enable getting logs for exited pods.

yifan-gu opened the pull request:
cc @dchen1107 @yujuhong

Commit:
kubelet/rkt: enable getting logs for exited pods.

dchen1107 commented:
Can you share with us some test result of kubectl logs with rkt here?

dchen1107 commented:
ok to test

yifan-gu commented:
Before this PR, if the pod exits, the logs are not fetchable. This fixes the problem by changing the parameters given to journalctl. (use `-m _HOSTNAME` instead of `-M`). See https://github.com/coreos/rkt/issues/1528 for more discussion on this.

yifan-gu commented:
Example outputs:
```
cluster/kubectl.sh logs loop 
Mon Oct  5 22:07:39 UTC 2015
Mon Oct  5 22:07:40 UTC 2015
Mon Oct  5 22:07:41 UTC 2015
Mon Oct  5 22:07:42 UTC 2015
Mon Oct  5 22:07:43 UTC 2015
Mon Oct  5 22:07:44 UTC 2015
Mon Oct  5 22:07:45 UTC 2015
```

```
cluster/kubectl.sh logs loop --since=2s
Mon Oct  5 22:08:36 UTC 2015
Mon Oct  5 22:08:37 UTC 2015
```

```
cluster/kubectl.sh logs loop cluster/kubectl.sh logs loop --tail=4
Mon Oct  5 22:07:39 UTC 2015
Mon Oct  5 22:07:40 UTC 2015
Mon Oct  5 22:07:41 UTC 2015
Mon Oct  5 22:07:42 UTC 2015
```

```
cluster/kubectl.sh logs loop cluster/kubectl.sh logs loop --tail=4
Mon Oct  5 22:07:39 UTC 2015
Mon Oct  5 22:07:40 UTC 2015
Mon Oct  5 22:07:41 UTC 2015
Mon Oct  5 22:07:42 UTC 2015
```

```
cluster/kubectl.sh logs loop --timestamps --tail=3
-- Logs begin at Mon 2015-09-28 17:00:07 PDT, end at Mon 2015-10-05 15:09:42 PDT. --
Oct 05 15:07:39 rkt-c9f5ae40-9c17-485f-b8f7-eb545cbd5c09 sh[4]: Mon Oct  5 22:07:39 UTC 2015
Oct 05 15:07:40 rkt-c9f5ae40-9c17-485f-b8f7-eb545cbd5c09 sh[4]: Mon Oct  5 22:07:40 UTC 2015
Oct 05 15:07:41 rkt-c9f5ae40-9c17-485f-b8f7-eb545cbd5c09 sh[4]: Mon Oct  5 22:07:41 UTC 2015
```

dchen1107 commented:
I understood this. If we have e2e setup to pick up this change automatically, I won't ask above question. But now we don't, at least Jenkins GCE e2e doesn't pick up this.

dchen1107 commented:
lgtm

yifan-gu commented:
@dchen1107 Yes, I am just going to run some e2e tests on this.

dchen1107 commented:
cc/ @runningwild

yifan-gu commented:
I just locally run 2 `Downward API` e2e tests which test the output of the pods:


```
...
Downward API volume 
  should provide labels and annotations files
...
STEP: Saw pod success
STEP: Trying to get logs from node 127.0.0.1 pod metadata-volume-9b411cfa-6bb6-11e5-abe9-28d244b00276 container client-container: <nil>
STEP: Successfully fetched pod logs:cluster="rack10"
builder="john-doe"
kubernetes.io/config.seen="2015-10-05T16:12:53.80772414-07:00"
kubernetes.io/config.source="api"metadata-volume-9b411cfa-6bb6-11e5-abe9-28d244b00276
...

Downward API 
  should provide pod name and namespace as env vars
...
STEP: Saw pod success
STEP: Trying to get logs from node 127.0.0.1 pod downward-api-b02dded2-6bb6-11e5-abe9-28d244b00276 container dapi-container: <nil>
STEP: Successfully fetched pod logs:KUBERNETES_PORT=tcp://10.0.0.1:443
KUBERNETES_SERVICE_PORT=443
USER=root
AC_APP_NAME=dapi-container
SHLVL=1
HOME=/root
LOGNAME=root
KUBERNETES_PORT_443_TCP_ADDR=10.0.0.1
POD_NAME=downward-api-b02dded2-6bb6-11e5-abe9-28d244b00276
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP_PROTO=tcp
SHELL=/bin/sh
KUBERNETES_PORT_443_TCP=tcp://10.0.0.1:443
POD_NAMESPACE=e2e-tests-downward-api-lp4om
PWD=/
KUBERNETES_SERVICE_HOST=10.0.0.1
AC_METADATA_URL=
...

Ran 2 of 185 Specs in 70.273 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 183 Skipped PASS
```

FWIW I edited the tests to use absolute path in the `container.Command[0]`, otherwise rkt will fail. This belongs to another issue I am going to fix.

yujuhong commented:
@yifan-gu, I am really looking forward to running e2e on jenkis with rkt. Let us know if there is anything we can help with :)

yifan-gu commented:
@yujuhong I don't know how to set it up the e2e jekins, maybe that's the thing we should discuss this week.
Though not all of them, but at least some of the e2e tests can be enabled for rkt now.

yujuhong commented:
> @yujuhong I don't know how to set it up the e2e jekins, maybe that's the thing we should discuss this week.

We should. I think it'd be very helpful having e2e verifying this.


> Though not all of them, but at least some of the e2e tests can be enabled for rkt now.

Could you come up with a regex (`ginkgo.skip`) that would work for rkt? (examples in https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh).
We can trim down the list over time.

yifan-gu commented:
@yujuhong Cool, will do.

Commit:
kubelet/rkt: fetch journal logs in json format.

This enables more fine-grained control over the things we want to
output. Also by closing the stdout/stderr of the journalctl process
when user hits `Ctrl-C` after `kubectl logs $POD -f`, this enables
the journalctl process to exit.

yifan-gu commented:
Added one commit to eliminate the information like _HOSTNAME(rkt-xxx) _PID in the log.
Now the output is like:
```
$ cluster/kubectl.sh logs loop 
hello 0
hello 1
hello 2
hello 3
hello 4
hello 5
hello 6
hello 7
hello 8
hello 9
hello 10
hello 11
hello 12
```

And with `--timestamps`
```
cluster/kubectl.sh logs loop --timestamps
2015-10-06 17:06:04 -0700 PDT hello 0
2015-10-06 17:06:05 -0700 PDT hello 1
2015-10-06 17:06:06 -0700 PDT hello 2
2015-10-06 17:06:07 -0700 PDT hello 3
2015-10-06 17:06:08 -0700 PDT hello 4
2015-10-06 17:06:09 -0700 PDT hello 5
2015-10-06 17:06:10 -0700 PDT hello 6
2015-10-06 17:06:11 -0700 PDT hello 7
2015-10-06 17:06:12 -0700 PDT hello 8
2015-10-06 17:06:13 -0700 PDT hello 9
2015-10-06 17:06:14 -0700 PDT hello 10
2015-10-06 17:06:15 -0700 PDT hello 11
2015-10-06 17:06:16 -0700 PDT hello 12
2015-10-06 17:06:17 -0700 PDT hello 13
2015-10-06 17:06:18 -0700 PDT hello 14
2015-10-06 17:06:19 -0700 PDT hello 15
2015-10-06 17:06:20 -0700 PDT hello 16
```

yifan-gu commented:
/cc @jonboulle If you have some time to give a look.

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+
+		ts, err := strconv.ParseInt(tss, 0, 64)
+		if err != nil {
+			glog.Warningf("rkt: Cannot parse the timestamp of the journal log, skipping the line: %v", err)
```
s/the//g

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+		b := scanner.Bytes()
+
+		if err := json.Unmarshal(b, &data); err != nil {
+			glog.Warningf("rkt: Cannot unmarshal the journal log, skipping the line: %v", err)
```
s/the//g

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+		// We need to close the reader to force the journalctl process to exit, also we need to
+		// return here to avoid goroutine leak.
+		if _, err := io.WriteString(w, result); err != nil {
+			glog.Warningf("rkt: Cannot write the log to output: %v, data: %s", err, result)
```
s/the//g

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+	go feedLog(&wg, logOptions, outPipe, stdout)
+	go feedLog(&wg, logOptions, errPipe, stderr)
+
+	// Wait until the logs are feeded to stdout, stderr.
```
s/feeded/fed/ (sorry, English is awfully inconsistent)

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+
+// feedLog reads and parses the journal json object from r,
+// and writes the logs line by line to w.
+func feedLog(wg *sync.WaitGroup, logOptions *api.PodLogOptions, r io.ReadCloser, w io.Writer) {
```
or perhaps just pipeLog?

jonboulle commented on pkg/kubelet/rkt/log.go:
```diff
@@ -0,0 +1,163 @@
+		since = logOptions.SinceTime.Unix()
+	}
+
+	cmd.Args = append(cmd.Args, "--since", time.Unix(since, 0).Format(journalSinceLayout))
```
is it always guaranteed that either SinceSeconds or SinceTime is set?

yifan-gu replied:
Nope, but I assumed `--since=0` is ok, however then I found https://github.com/systemd/systemd/issues/1477

jonboulle replied:
@yifan-gu what's the plan here?

yifan-gu replied:
@jonboulle Guess just workaround with `-r` for the time being. And update this if https://github.com/systemd/systemd/issues/1477 gets fixed ever.

dchen1107 commented:
@yifan-gu Please rebase. 

According to run e2e tests against rkt, I filed an issue a while back: https://github.com/kubernetes/kubernetes/issues/7205

yifan-gu commented:
Rebased.

jonboulle commented:
LGTM

On Tue, Oct 6, 2015 at 5:16 PM, Yifan Gu <notifications@github.com> wrote:

> /cc @jonboulle <https://github.com/jonboulle> If you have some time to
> give a look.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15114#issuecomment-146041832>
> .
>

yifan-gu commented:
ping?

yifan-gu commented:
ping on this?

yujuhong commented:
LGTM.
